### PR TITLE
support external queue implementations

### PR DESCRIPTION
**Pull request by psychocoderHPC**
solves #820

By providing a queue interface class it is possible to handle user defines queues within alpaka.

- add queue interface `ICpuQueue`
- remove usage of hardcoded register function for each queue type within the device

The second commit provides a OpenMP collective queue as test case for an externally defined queue which is designed with the requirements of #820 in mind.

### CODE REVIEW

**Comment by psychocoderHPC** [comment]
CC-ing: @krzikalla

**Comment by psychocoderHPC** [comment]
As I wrote in #820 I do not think that the highly specific collective queue implementation fitts into the generic concepts for alpaka. So IMO this is the way how to support the user requirements.

File: test/unit/queue/src/QueueTest.cpp [hunk 1]
```diff
@@ -159,7 +161,27 @@ void operator()()
...
+        ,
+        std::tuple<alpaka::dev::DevCudaRt, alpaka::queue::QueueCudaRtBlocking>,
+        std::tuple<alpaka::dev::DevCudaRt, alpaka::queue::QueueCudaRtNonBlocking>
+#endif
+#ifdef ALPAKA_ACC_GPU_HIP_ENABLED
+        ,
+        std::tuple<alpaka::dev::DevHipRt, alpaka::queue::QueueHipRtBlocking>,
+        std::tuple<alpaka::dev::DevHipRt, alpaka::queue::QueueHipRtNonBlocking>
+#endif
+    >;
```
Comment (BenjaminW3) [inline test/unit/queue/src/QueueTest.cpp:171]: Why did you remove it completely from `test/queue/Queue.hpp`? I would prefer to still have the `TestQueues` there and only appending the `QueueCpuOmp2Collective` here.

Comment (BenjaminW3) [reply test/unit/queue/src/QueueTest.cpp:171]: Or simply add it to `TestQueues` there.

File: test/unit/event/src/EventTest.cpp [hunk 2]
```diff
@@ -285,7 +286,28 @@ void operator()()
...
+        ,
+        std::tuple<alpaka::dev::DevCudaRt, alpaka::queue::QueueCudaRtBlocking>,
+        std::tuple<alpaka::dev::DevCudaRt, alpaka::queue::QueueCudaRtNonBlocking>
+#endif
+#ifdef ALPAKA_ACC_GPU_HIP_ENABLED
+        ,
+        std::tuple<alpaka::dev::DevHipRt, alpaka::queue::QueueHipRtBlocking>,
+        std::tuple<alpaka::dev::DevHipRt, alpaka::queue::QueueHipRtNonBlocking>
+#endif
+    >;
```
Comment (BenjaminW3) [inline test/unit/event/src/EventTest.cpp:295]: Same as in `QueueTest.cpp`

File: include/alpaka/queue/cpu/ICpuQueue.hpp [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+            //#############################################################################
+            //! The CPU queue interface
+            class ICpuQueue
+            {
+            public:
+                //-----------------------------------------------------------------------------
+                ICpuQueue() = default;
+                //! enqueue the event into the given queue ------------------------------------
+                virtual void enqueue(std::shared_ptr<ICpuQueue> &, event::EventCpu &) = 0;
+                //! the given queue is waiting for the event-----------------------------------
```
Comment (BenjaminW3) [inline include/alpaka/queue/cpu/ICpuQueue.hpp]: Could you please use the current style with a separate line for the comment. This would make it easier to mass-replace them later.

File: include/alpaka/queue/cpu/ICpuQueue.hpp [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+ *
+ * This file is part of Alpaka.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+
+#pragma once
```
Comment (BenjaminW3) [inline include/alpaka/queue/cpu/ICpuQueue.hpp:10]: Could you please remove the second new-line?

File: test/unit/queue/src/QueueTest.cpp (hunk 1 above)
Comment (psychocoderHPC) [reply test/unit/queue/src/QueueTest.cpp:171]: I moved it to the tests because in this file defines traits like `IsBlockingQueue`. This make it impossible to have all trait definitions outside of the file `Queue.hpp` but I do not like to define the traits within `Queue.hpp` because I like to have clear borders between the "user code` (the collective queue) and alpaka.

File: include/alpaka/queue/cpu/ICpuQueue.hpp (hunk 3 above)
Comment (psychocoderHPC) [reply include/alpaka/queue/cpu/ICpuQueue.hpp]: ^^ I do not find examples for the style but think I got what you mean

File: include/alpaka/queue/QueueCpuNonBlocking.hpp [hunk 5]
```diff
@@ -73,6 +74,20 @@ namespace alpaka
                     //-----------------------------------------------------------------------------
                     ~QueueCpuNonBlockingImpl() = default;
 
+                    //-----------------------------------------------------------------------------
+                    void enqueue(std::shared_ptr<cpu::ICpuQueue> & iQueue, event::EventCpu & ev) final
+                    {
+                        auto spQueueImpl = std::static_pointer_cast<QueueCpuNonBlockingImpl>(iQueue);
```
Comment (BenjaminW3) [inline include/alpaka/queue/QueueCpuNonBlocking.hpp]: Do we really need the separate `iQueue` parameter? We should have a `enqueue` and `wait` overload for the non-shared_ptr types. Then we only would have to pass `*this`.

File: test/unit/queue/src/QueueTest.cpp (hunk 1 above)
Comment (BenjaminW3) [reply test/unit/queue/src/QueueTest.cpp:171]: But then please leave the alpaka-queues in there and append the new one.

**Comment by BenjaminW3** [comment]
There is another option to the interface inheritance. We could probably give the `DevCpu` a template parameter for additional queues it has to support. It then would hold a vector of variants of the built-in queues and the new queues. However, I do not think that this would be better.

File: include/alpaka/queue/QueueCpuBlocking.hpp [hunk 6]
```diff
@@ -217,3 +232,5 @@ namespace alpaka
         }
     }
 }
+
+#include <alpaka/event/EventCpu.hpp>
```
Comment (BenjaminW3) [inline include/alpaka/queue/QueueCpuBlocking.hpp:241]: Why is this strange include necessary?

File: test/unit/queue/src/CollectiveQueue.cpp [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+    // only one thread is allowed to flip the value
+    REQUIRE(numFlippedValues == 1u);
+    // only one value is allowed to differ from the initial value
+    REQUIRE(numNonIntitialValues == 1u);
+}
+};
+
+TEST_CASE( "queueCollective", "[queue]")
+{
+    TestCollective{}();
```
Comment (BenjaminW3) [inline test/unit/queue/src/CollectiveQueue.cpp]: Do we need the separate class at all? It is only used in the templatized tests as workaround. We should be able to move the code directly in here.

File: test/common/include/alpaka/test/queue/QueueCpuOmp2Collective.hpp [hunk 8]
```diff
@@ -0,0 +1,454 @@
+/* Copyright 2019 Axel Huebl, Benjamin Worpitz, Matthias Werner
+ *
+ * This file is part of Alpaka.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+
```
Comment (BenjaminW3) [inline test/common/include/alpaka/test/queue/QueueCpuOmp2Collective.hpp]: Please remove the second newline here as well.

File: test/common/include/alpaka/test/queue/QueueCpuOmp2Collective.hpp [hunk 9]
```diff
@@ -0,0 +1,454 @@
...
+#pragma once
+
+#ifdef ALPAKA_ACC_CPU_B_OMP2_T_SEQ_ENABLED
+
```
Comment (BenjaminW3) [inline test/common/include/alpaka/test/queue/QueueCpuOmp2Collective.hpp:13]: Do we need the check here similar to all other OMP dependant files?
```
#if _OPENMP < 200203
    #error If ALPAKA_ACC_CPU_B_SEQ_T_OMP2_ENABLED is set, the compiler has to support OpenMP 2.0 or higher!
#endif
```

File: test/common/include/alpaka/test/queue/QueueCpuOmp2Collective.hpp [hunk 10]
```diff
@@ -0,0 +1,454 @@
...
+        //-----------------------------------------------------------------------------
+        //! The test queue specifics.
+        namespace queue
+        {
+            namespace traits
+            {
+                //#############################################################################
+                //! The blocking queue trait specialization for a OpenMP2 collective CPU queue.
+                template<>
+                struct IsBlockingQueue<
```
Comment (BenjaminW3) [inline test/common/include/alpaka/test/queue/QueueCpuOmp2Collective.hpp:450]: Could we move the `IsBlockingQueue` trait (maybe as `IsBlocking`) for this and all the other queues to the main `alpaka` library?
Could we please rename this queue to `QueueCpuBlockingOmp2Collective` so hat it matches the style of all others?

File: include/alpaka/queue/QueueCpuBlocking.hpp (hunk 6 above)
Comment (psychocoderHPC) [reply include/alpaka/queue/QueueCpuBlocking.hpp:241]: This is needed to pass the header checks where we check that a header is including all dependencies e.g. https://github.com/ComputationalRadiationPhysics/alpaka/pull/822/files#diff-5b4348d85ae1c4f4c1c432f2aa3b97bdR74 will call wait with an event but the specialization for the meaning of wait is within `EventCpu.hpp`. But since event depends on queue and here we depend of knowing the event we need to pull in the specializations.

File: test/common/include/alpaka/test/queue/QueueCpuOmp2Collective.hpp (hunk 10 above)
Comment (psychocoderHPC) [reply test/common/include/alpaka/test/queue/QueueCpuOmp2Collective.hpp:450]: > Could we please rename this queue to `QueueCpuBlockingOmp2Collective` so that it matches the style of all others?

I also thought at this name before but it is not blocking within a parallel region so I avoided to use blocking within the name

Comment (psychocoderHPC) [reply test/common/include/alpaka/test/queue/QueueCpuOmp2Collective.hpp:450]: > Could we move the `IsBlockingQueue` trait (maybe as `IsBlocking`) for this and all the other queues to the main `alpaka` library?

We can do it but outside of this PR.

Comment (BenjaminW3) [reply test/common/include/alpaka/test/queue/QueueCpuOmp2Collective.hpp:450]: But why does `IsBlockingQueue` then return true? A queue that is blocking is defined by the fact that a call to `queue::enqueue(queue, task);` (which is also hidden within `kernel::exec`) does not return before the task is executed. Is this true or false for this queue?

File: include/alpaka/queue/QueueCpuNonBlocking.hpp (hunk 5 above)
Comment (psychocoderHPC) [reply include/alpaka/queue/QueueCpuNonBlocking.hpp]: I will try it again. I started with that before but somehow I was unhappy with that. I do not know anymore why so I will try it again.

File: test/common/include/alpaka/test/queue/QueueCpuOmp2Collective.hpp (hunk 10 above)
Comment (psychocoderHPC) [reply test/common/include/alpaka/test/queue/QueueCpuOmp2Collective.hpp:450]: because it depends where you use it (in a parallel region or not) outside it is blocking and to not move over tests those require a nonblocking queue it is set to blocking.

File: include/alpaka/queue/cpu/ICpuQueue.hpp [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+        namespace cpu
+        {
+
+            //#############################################################################
+            //! The CPU queue interface
+            class ICpuQueue
+            {
+            public:
+                //-----------------------------------------------------------------------------
+                ICpuQueue() = default;
```
Comment (BenjaminW3) [inline include/alpaka/queue/cpu/ICpuQueue.hpp]: Do we really need the explicitly defaulted default construcor? I think this is not necessary for an abstract interface class.

File: include/alpaka/queue/QueueCpuNonBlocking.hpp [hunk 12]
```diff
@@ -73,6 +74,18 @@ namespace alpaka
                     //-----------------------------------------------------------------------------
                     ~QueueCpuNonBlockingImpl() = default;
```
Comment (BenjaminW3) [inline include/alpaka/queue/QueueCpuNonBlocking.hpp]: We should now remove this destructor as it is missing the virtual keyword and it now has a base class.

File: include/alpaka/queue/QueueCpuBlocking.hpp [hunk 13]
```diff
@@ -59,6 +60,18 @@ namespace alpaka
                     //-----------------------------------------------------------------------------
                     ~QueueCpuBlockingImpl() = default;
```
Comment (BenjaminW3) [inline include/alpaka/queue/QueueCpuBlocking.hpp]: We should now remove this destructor as it is missing the virtual keyword and it now has a base class.

File: test/common/include/alpaka/test/queue/QueueCpuOmp2Collective.hpp [hunk 14]
```diff
@@ -0,0 +1,455 @@
...
+                    //-----------------------------------------------------------------------------
+                    QueueCpuOmp2CollectiveImpl(QueueCpuOmp2CollectiveImpl const &) = delete;
+                    //-----------------------------------------------------------------------------
+                    QueueCpuOmp2CollectiveImpl(QueueCpuOmp2CollectiveImpl &&) = delete;
+                    //-----------------------------------------------------------------------------
+                    auto operator=(QueueCpuOmp2CollectiveImpl const &) -> QueueCpuOmp2CollectiveImpl & = delete;
+                    //-----------------------------------------------------------------------------
+                    auto operator=(QueueCpuOmp2CollectiveImpl &&) -> QueueCpuOmp2CollectiveImpl & = delete;
+                    //-----------------------------------------------------------------------------
+                    ~QueueCpuOmp2CollectiveImpl() = default;
```
Comment (BenjaminW3) [inline test/common/include/alpaka/test/queue/QueueCpuOmp2Collective.hpp]: Here as well

**Comment by BenjaminW3** [comment]
CI complains: `CollectiveQueue.cpp(57): error: __host__ or __device__ annotation on lambda requires --expt-extended-lambda nvcc flag`

**Review by BenjaminW3 — APPROVED** [review]
